lib/rust/flexer/tests/flexer_lexer_definition_test.rs

```diff
@@ -0,0 +1,229 @@
+//! This file contains tests for the intended defnition code of a lexer using the flexer, based on
+//! the following small language.
+//!
+//! The language here is being defined as follows:
+//!
+//! a-word      = 'a'+;
+//! b-word      = 'b'+;
+//! word        = a-word | b-word;
+//! space       = ' ';
+//! spaced-word = space, word;
+//! language    = word, spaced-word*;
+
+use flexer::prelude::*;
+use flexer::automata::dfa::DFA;
+use flexer::automata::nfa::NFA;
+use flexer::group::Group;
+use lazy_reader::{BookmarkId, LazyReader, Reader};
+use flexer::{FlexerState, Flexer};
+use lazy_reader::decoder::DecoderUTF8;
+use flexer::automata::pattern::Pattern;
+
+
+// ===========
+// === AST ===
+// ===========
+
+/// A very simple AST, sufficient for the simple lexer being defined.
+#[derive(Clone,Debug,PartialEq)]
+pub enum AST {
+    /// A word from the input, consisting of a sequence of all `a` or all `b`.
+    Word(String),
+    /// A token that the lexer is unable to recognise.
+    Unrecognised(String)
+}
+
+
+
+// =============
+// === Lexer ===
+// =============
+
+#[allow(missing_docs)]
+#[derive(Clone,Debug)]
+pub struct Lexer {
+    groups: Vec<Group>
+}
+
+#[allow(missing_docs)]
+impl Lexer {
+    pub fn new() -> Self {
+        let groups = Vec::new();
+        Lexer{groups}
+    }
+
+    // TODO [AA] Parent groups.
+    pub fn define_group(&mut self,name:&str) -> &mut Group {
+        let id    = self.groups.len();
+        let group = Group::new(id,String::from(name),None);
+        self.groups.push(group);
+        self.groups.get_mut(id).expect("Has just been pushed so should always exist.")
```

Comment:
this is ugly. Please add to prelude method to Vec like `push_and_get_mut` instead.

Reply:
How can I do this? We don't own `Vec`.

Reply:
Note to self, see `option.rs`.